### ISSUE TITLE
Investigate performance of rest parameter indice optimisation 

### ISSUE THREAD

**Issue by sebmck** [babel/babel#1845]
The `arguments` rest parameter optimisation is actually a deoptimisation.

```javascript
function fetch() {
  return arguments[1];
}
```

is slower than

```javascript
function allocate() {
  for (var _len = arguments.length, args = Array(_len > 1 ? _len - 1 : 0), _key = 1; _key < _len; _key++) {
    args[_key - 1] = arguments[_key];
  }

  return args[0];
}
```

Benchmark:

```javascript
var Benchmark = require("benchmark");
var suite = new Benchmark.Suite;

function fetch() {
  return arguments[1];
}

function allocate() {
  for (var _len = arguments.length, args = Array(_len > 1 ? _len - 1 : 0), _key = 1; _key < _len; _key++) {
    args[_key - 1] = arguments[_key];
  }

  return args[0];
}

suite.add("allocate", function() {
  allocate();
});

suite.add("fetch", function() {
  fetch();
});

suite.on("cycle", function(event) {
  console.log(String(event.target));
})

suite.run();
```

cc @stefanpenner @benjamingr

**Comment by benjamingr** [comment]
```
Benjamins-Macbook-Air:tmp Benjamin$ node benchme.js 
allocate x 20,875,118 ops/sec ±3.15% (81 runs sampled)
fetch x 20,351,523 ops/sec ±2.22% (85 runs sampled)
```

Are we sure it's not just DCE? (dead code elimination)

This is the result with anti DCE code in place:

```
Benjamins-Macbook-Air:tmp Benjamin$ node benchme.js 
allocate x 6,045,442 ops/sec ±12.25% (61 runs sampled)
fetch x 15,229,668 ops/sec ±4.60% (75 runs sampled)
```

What I did was change fetch and allocate to:

```js
function fetch() {
  return arguments[1];
  eval("debugger;");
}

function allocate() {
  for (var _len = arguments.length, args = Array(_len > 1 ? _len - 1 : 0), _key = 1; _key < _len; _key++) {
    args[_key - 1] = arguments[_key];
  }

  return args[0];
  eval("debugger;");
}
```

To force DCE not to run.

**Comment by sebmck** [comment]
Ah right, completely forgot about that. Thanks for looking into it @benjamingr! Looks like this really needs to be addressed in [six-speed](https://github.com/kpdecker/six-speed) then since it's being passed around a lot recently.

**Comment by benjamingr** [comment]
@sebmck performance is extremely hard in JavaScript, trust nothing :D 

My first go-to is to check for deopts and look at assembly, node/io will let you do that (and the d8 shell is even better at this) - `--trace-deops` is invaluable sometimes and petka started working on https://github.com/petkaantonov/nodeperf but sort of abandoned it.

**Comment by sebmck** [comment]
cc @kpdecker

**Comment by stefanpenner** [comment]
one thing to be careful about is that `arguments[outOfBoundsIndex]` is a slight perf hit.

**Comment by benjamingr** [comment]
@stefanpenner `arguments[outOfBoundsIndex]` is extremely rare, I don't think it's worth optimizing for to be completely honest.

**Comment by sebmck** [comment]
@Mark-Simulacrum did some tests for default parameters and had some interesting results:

```
out-of-bounds access x 14,918,297 ops/sec ±0.26% (100 runs sampled)
in-bounds access x 40,865,466 ops/sec ±0.26% (102 runs sampled)
out-of-bounds access w/ check x 41,564,831 ops/sec ±0.17% (102 runs sampled)
in-bounds access w/ check x 32,727,436 ops/sec ±0.30% (99 runs sampled)
```

out-of-bounds were made faster but in-bounds were made slower. Tradeoffs for either.

**Comment by benjamingr** [comment]
Can I see the benchmarks?

**Comment by sebmck** [comment]
Don't think @Mark-Simulacrum posted it.

**Comment by sebmck** [comment]
Actually, it's [here](https://gist.github.com/sebmck/97e1bb25569e6ea459d4). DCE shouldn't be an issue either:

```
ooflorent [2:14 PM]
we must check dead-code elimination

ooflorent [2:15 PM]
well, i'll try it

ooflorent [2:16 PM]2:16
no DCE

ooflorent [2:16 PM]
cool :simple_smile:
```

**Comment by stefanpenner** [comment]
its also worth noting that function body size, does affect inlining potential. Which can hurt other things, we just need to be careful at looking at the whole picture, not just in isolation. 

silly perf :P

**Comment by RReverser** [comment]
Just so that other also see our private discussion - fixed the suite so that we wouldn't introduce nor measure external deopts but only actual arguments: https://gist.github.com/RReverser/69ff302cfdeda4d01dc4

```
out-of-bounds access x 3,644,966 ops/sec ±2.51% (86 runs sampled)
in-bounds access x 30,417,099 ops/sec ±8.36% (86 runs sampled)
out-of-bounds access w/ check x 65,782,176 ops/sec ±1.26% (85 runs sampled)
in-bounds access w/ check x 53,554,089 ops/sec ±4.67% (81 runs sampled)
```

**Comment by RReverser** [comment]
@sebmck Ping (just because I know this is a closed issue, so not sure whether you get notifications / what is the status).

**Comment by sebmck** [comment]
Can you verify that there is no DCE happening and that the same performance improvements happen on a wide range of devices (various Android browsers, iOS Safari, Firefox, Chrome, IE etc). There's not much proof in a single benchmark in a single environment.

**Comment by RReverser** [comment]
@sebmck We were talking precisely about V8 deopt, so other browsers don't matter here (they have pretty much same performance). I can guarantee that IRHydra doesn't show deopt for second case and does for the first one though.

**Comment by sebmck** [comment]
Not sure if doing engine specific optimisations is a good idea. This would be a good fit for a plugin though. It's likely you'll see worse performance on other browsers where this deoptimisation doesn't occur.

**Comment by RReverser** [comment]
Just checked on Chakra (IE/Edge):
```
out-of-bounds access x 3,551,416 ops/sec ±3.43% (85 runs sampled)
in-bounds access x 127,760,828 ops/sec ±2.95% (87 runs sampled)
out-of-bounds access w/ check x 118,674,682 ops/sec ±4.19% (85 runs sampled)
in-bounds access w/ check x 102,588,306 ops/sec ±2.25% (90 runs sampled)
```

**Comment by RReverser** [comment]
> Can you verify that there is no DCE happening

The entire idea of this optimization is to enforce DCE (so that separate fast paths are generated for (1) when extra arg is passed and (2) it's not).

**Comment by sebmck** [comment]
So in bound access is slower with the check then? It just seems pointless to worry about these kinds of microoptimisations when it varies so wildly.

**Comment by RReverser** [comment]
Yeah, a bit (although pretty much comparable). It's not that much about microoptimization of default parameter itself, but rather of entire function (macro improvement), because due to deopt not only line with default parameter is punished, but entire function is, which is sad - means any perf-critical function needs to use custom default parameters handler if it doesn't want to get completely deoptimized. This is also the case with current version of V8 (they don't have optimizations almost for any ES6 features yet), but can be easily solved on Babel side.

**Comment by RReverser** [comment]
However, if you feel this is out of scope for Babel, I'm totally fine with externalizing optimizations into a separate plugin.

**Comment by sebmck** [comment]
Thoughts @stefanpenner @benjamingr?

**Comment by benjamingr** [comment]
I'll have to give this a day to examine the benchmarks (although @RReverser 's bench looks legit)

Another point is that arguments access out of bounds sounds like a really really rare case anyway and people would expect `rest[i]` to behave like `arguments[i]`

**Comment by stefanpenner** [comment]
@RReverser findings mirror similar research I have conducted in the past.

In those tests, outcomes mirrored the above findings, even when run across a wide range of platforms. that is to say, they all are in-favor of the out-of-bounds check.

Being able to rely on babel's compiler  in these hot-spots would be rad, So I am :+1.

**Comment by RReverser** [comment]
> Another point is that arguments access out of bounds sounds like a really really rare case anyway

Not when used implicitly by default arguments. Any function like `(options = {}) => {}` currently causes access out of bounds in generated code.

**Comment by sebmck** [comment]
Great, can you create another issue @RReverser tracking this with suggested transform changes.

**Comment by benjamingr** [comment]
How would `(options = {}) => {} ` generate an out of bounds check?

I thought we were just talking about converting rest parameters to
`arguments`, when default values are involved this definitely shouldn't be
an out of bounds check :)

**Comment by sebmck** [comment]
The conversation shifted to default parameters and it's output, ie:

```javascript
function foo(bar = "") {}
```

```javascript
function foo() {
  var bar = arguments[0] === undefined ? "" : arguments[0];
}
```

**Comment by benjamingr** [comment]
Oh ok, sorry, as mentioned above I didn't really have time to go through
everything yet :)

Isn't the straightforward way to convert

```js
function(a = 5, b = 3, c = {}){

}
```

into:

```js
function(a, b, c){
     switch(arguments.length) { // not a deopt afaik
       case 0: a = 5;
       case 1: b = 3;
       case 2: c = {};
     }
}
```

Or am I missing something obvious?

On Sun, Jul 5, 2015 at 2:28 AM, Sebastian McKenzie <notifications@github.com

**Comment by benjamingr** [comment]
Oh nevermind, I am. Just remembered `undefined` in default arguments
triggers the default argument. My bad.

**Comment by kpdecker** [comment]
Not helpful for existing browsers, but the FF folks have been talking about this issue: https://bugzilla.mozilla.org/show_bug.cgi?id=1131996

**Comment by RReverser** [comment]
> Oh nevermind, I am. Just remembered `undefined` in default arguments triggers the default argument. My bad.

Also, we can't emit argument names in the case of defaults as it changes function's `.length`.

**Comment by sebmck** [comment]
Addressed out of bounds in default parameters in 0509fb148f71e0cceaf60a7f35ab1a96091e1ca2.
